packages/dds/tree/src/feature-libraries/sequence-field/utils.ts

```diff
@@ -301,3 +381,396 @@ export function tryExtendMark(lhs: ObjectMark, rhs: Readonly<ObjectMark>): boole
+        } else {
+            const splitMark: MoveOut<T> | ReturnFrom<T> = {
+                ...mark,
+                id: part.id,
+                count: part.count ?? mark.count,
+            };
+            if (part.modifyAfter !== undefined) {
+                assert(
+                    composeChildren !== undefined,
+                    "Must provide a change composer if modifying moves",
+                );
+                const changes = composeChildren(mark.changes, part.modifyAfter);
+                if (changes !== undefined) {
+                    splitMark.changes = changes;
+                } else {
+                    delete splitMark.changes;
+                }
+            }
+            result.push(splitMark);
+        }
+    }
+    return result;
+}
+
+export function applyMoveEffectsToMark<T>(
+    inputMark: Mark<T>,
+    revision: RevisionTag | undefined,
+    moveEffects: MoveEffectTable<T>,
+    genId: IdAllocator,
+    reassignIds: boolean,
+    composeChildren?: (a: T | undefined, b: T | undefined) => T | undefined,
+): Mark<T>[] {
+    let mark = inputMark;
+    if (isMoveMark(mark)) {
+        if (reassignIds) {
+            const newId = getUniqueMoveId(mark, revision, genId, moveEffects);
+            if (newId !== mark.id) {
+                mark = clone(mark);
+                mark.id = newId;
+                moveEffects.validatedMarks.add(mark);
+            }
+        }
+
+        const type = mark.type;
+        switch (type) {
+            case "MoveOut":
+            case "ReturnFrom": {
+                const effect = moveEffects.srcEffects.get(mark.id);
+                if (effect !== undefined) {
+                    moveEffects.srcEffects.delete(mark.id);
+                    const splitMarks = splitMoveOut(mark, effect, composeChildren);
+                    for (const splitMark of splitMarks) {
+                        moveEffects.validatedMarks.add(splitMark);
+                    }
+                    return splitMarks;
+                }
+                break;
+            }
+            case "MoveIn":
+            case "ReturnTo": {
+                const effect = moveEffects.dstEffects.get(mark.id);
+                if (effect !== undefined) {
+                    moveEffects.dstEffects.delete(mark.id);
+                    const splitMarks = splitMoveIn(mark, effect);
+                    for (const splitMark of splitMarks) {
+                        moveEffects.validatedMarks.add(splitMark);
+                    }
+                    return splitMarks;
+                }
+                break;
+            }
+            default:
+                unreachableCase(type);
+        }
+    }
+    return [mark];
+}
+
+export function getUniqueMoveId<T>(
+    mark: MoveMark<T>,
+    revision: RevisionTag | undefined,
+    genId: IdAllocator,
+    moveEffects: MoveEffectTable<T>,
+): MoveId {
+    if (!moveEffects.validatedMarks.has(mark) && (mark.revision ?? revision === undefined)) {
```

Comment:
The operator === has higher precendence than operator ??, so `mark.revision ?? revision === undefined` is equivalent to `mark.revision ?? (revision === undefined)`, which is probably not want you wanted here.